src/rpc/rawtransaction.cpp

```diff
@@ -924,10 +924,19 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
 
     TxValidationState state;
     bool test_accept_res;
+    CAmount fee{0}; // To return transaction fee from AcceptToMemoryPool
     {
         LOCK(cs_main);
         test_accept_res = AcceptToMemoryPool(mempool, state, std::move(tx),
-            nullptr /* plTxnReplaced */, false /* bypass_limits */, max_raw_tx_fee, /* test_accept */ true);
+            nullptr /* plTxnReplaced */, false /* bypass_limits */, /* test_accept */ true, &fee);
+    }
+
+    // Check that fee does not exceed maxfee
+    if (test_accept_res && max_raw_tx_fee && fee > max_raw_tx_fee) {
+        result_0.pushKV("allowed", false);
+        result_0.pushKV("reject-reason", "max-fee-exceeded");
+        result.push_back(std::move(result_0));
+        return result;
```

Comment:
ea8eb397 and 3721417 suggestion

`git diff -w`
```diff
-    // Check that fee does not exceed maxfee
     if (test_accept_res && max_raw_tx_fee && fee > max_raw_tx_fee) {
+        // Accepted, but disallow because fee exceeds maxfee
         result_0.pushKV("allowed", false);
         result_0.pushKV("reject-reason", "max-fee-exceeded");
-        result.push_back(std::move(result_0));
-        return result;
-    }
+    } else {
         result_0.pushKV("allowed", test_accept_res);
@@ -950,6 +948,7 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
                 result_0.pushKV("reject-reason", state.GetRejectReason());
             }
         }
+    }
 
     result.push_back(std::move(result_0));
     return result;
```

<details><summary>code</summary><p>

```cpp
    if (test_accept_res && max_raw_tx_fee && fee > max_raw_tx_fee) {
        // Accepted, but disallow because fee exceeds maxfee
        result_0.pushKV("allowed", false);
        result_0.pushKV("reject-reason", "max-fee-exceeded");
    } else {
        result_0.pushKV("allowed", test_accept_res);
        if (!test_accept_res) {
            if (state.IsInvalid()) {
                if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
                    result_0.pushKV("reject-reason", "missing-inputs");
                } else {
                    result_0.pushKV("reject-reason", strprintf("%s", state.GetRejectReason()));
                }
            } else {
                result_0.pushKV("reject-reason", state.GetRejectReason());
            }
        }
    }

    result.push_back(std::move(result_0));
    return result;
}
```
</p></details>

Reply:
Both seem fine. I have a preference for how it is right now to avoid the deeply nested if statements.

Reply:
Absurdfee is kind of a "special case" where `test_accept_res` is true, but we still say not allowed... maybe it's good to keep it in a separate block?

Reply:
I think (a) keep processing separate from move-and-return, which is why it is separated by an empty line at the end of the function, (b) don't repeat yourself, and (c) the principle of least surprise (at least, I found it surprising) outweigh the also-valid (d) avoid nested conditionals, but it's a detail that I won't hold against anyone :smiling_face_with_three_hearts: 

OTOH, I find a bit odd the logic of `test_accept_res` being true but being possibly disallowed due to the maxfee. Perhaps rename it (it's also not named like a boolean, e.g. `is/has/verb_{some_state}`, or mention it in the code like my doc suggestion (which could be improved). I think it's good to optimise the overall section for clarity and coherence, even if it changes a few more lines of code.

But these are all code style opinions, feel free to ignore if you disagree.

Reply:
Another suggestion (but it's also good as-is):
<details>
<summary>Diff</summary>

```diff
--- a/src/rpc/rawtransaction.cpp
+++ b/src/rpc/rawtransaction.cpp
@@ -923,34 +923,36 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
     result_0.pushKV("txid", tx_hash.GetHex());
 
     TxValidationState state;
-    bool test_accept_res;
+    bool allowed;
     CAmount fee{0}; // To return transaction fee from AcceptToMemoryPool
     {
         LOCK(cs_main);
-        test_accept_res = AcceptToMemoryPool(mempool, state, std::move(tx),
+        allowed = AcceptToMemoryPool(mempool, state, std::move(tx),
             nullptr /* plTxnReplaced */, false /* bypass_limits */, /* test_accept */ true, &fee);
     }
+    std::string reject_reason;
 
     // Check that fee does not exceed maxfee
-    if (test_accept_res && max_raw_tx_fee && fee > max_raw_tx_fee) {
-        result_0.pushKV("allowed", false);
-        result_0.pushKV("reject-reason", "max-fee-exceeded");
-        result.push_back(std::move(result_0));
-        return result;
+    if (allowed) {
+        if (max_raw_tx_fee && fee > max_raw_tx_fee) {
+            allowed = false;
+            reject_reason = "max-fee-exceeded";
         }
-    result_0.pushKV("allowed", test_accept_res);
-    if (!test_accept_res) {
+    } else {
         if (state.IsInvalid()) {
             if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
-                result_0.pushKV("reject-reason", "missing-inputs");
+                reject_reason = "missing-inputs";
             } else {
-                result_0.pushKV("reject-reason", strprintf("%s", state.GetRejectReason()));
+                reject_reason = strprintf("%s", state.GetRejectReason());
             }
         } else {
-            result_0.pushKV("reject-reason", state.GetRejectReason());
+            reject_reason = state.GetRejectReason();
         }
     }
-
+    result_0.pushKV("allowed", allowed);
+    if (!reject_reason.empty()) {
+        result_0.pushKV("reject-reason", reject_reason);
+    }
     result.push_back(std::move(result_0));
     return result;
 }
```
</details>
<details>
<summary>Code</summary>

```
    bool allowed;
    CAmount fee{0}; // To return transaction fee from AcceptToMemoryPool
    {
        LOCK(cs_main);
        allowed = AcceptToMemoryPool(mempool, state, std::move(tx),
            nullptr /* plTxnReplaced */, false /* bypass_limits */, /* test_accept */ true, &fee);
    }
    std::string reject_reason;

    // Check that fee does not exceed maxfee
    if (allowed) {
        if (max_raw_tx_fee && fee > max_raw_tx_fee) {
            allowed = false;
            reject_reason = "max-fee-exceeded";
        }
    } else {
        if (state.IsInvalid()) {
            if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
                reject_reason = "missing-inputs";
            } else {
                reject_reason = strprintf("%s", state.GetRejectReason());
            }
        } else {
            reject_reason = state.GetRejectReason();
        }
    }
    result_0.pushKV("allowed", allowed);
    if (!reject_reason.empty()) {
        result_0.pushKV("reject-reason", reject_reason);
    }
    result.push_back(std::move(result_0));
    return result;
```

<details>